### PR TITLE
Pull out segment flush threshold computation to make it easier to test

### PR DESCRIPTION
**Pull request by mneedham**
## Description

I wanted to understand how the segment-based flush threshold updater works, so I pulled out the computation bit of code so that I could write some tests to see how different inputs affect the threshold. 

I haven't changed the way it works - only pulled out code from `SegmentSizeBasedFlushThresholdUpdater` into  `SegmentFlushThresholdComputer`.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/Clock.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller.helix.core.realtime.segment;
+
+interface Clock {
```
Comment (richardstartin) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/Clock.java:21]: We can use `java.time.Clock` which has the `millis()` method instead. It has a system clock at any offset and also allows use of a fixed clock for testing.

**Comment by richardstartin** [comment]
Injectable clocks are essential if you want to test time-dependent behaviour, great initiative 👍🏻

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    double currentRatio = (double) numRowsConsumed / committingSegmentSizeBytes;
+    // Compute segment size to rows ratio only from the lowest available partition id.
+    // If we consider all partitions then it is likely that we will assign a much higher weight to the most
+    // recent trend in the table (since it is usually true that all partitions of the same table follow more or
+    // less same characteristics at any one point in time).
+    // However, when we start a new table or change controller mastership, we can have any partition completing first.
+    // It is best to learn the ratio as quickly as we can, so we allow any partition to supply the value.
+    int smallestAvailablePartitionGroupId =
+        partitionGroupMetadataList.stream().mapToInt(PartitionGroupMetadata::getPartitionGroupId).min()
+            .orElseGet(() -> 0);
```
Comment (richardstartin) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java:110]: I know this is just code moved from one place to another but `.orElse(0)` would be better

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/Clock.java (hunk 1 above)
Comment (mneedham) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/Clock.java:21]: Oh nice! I didn't know you could do that, will update.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java (hunk 2 above)
Comment (mneedham) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java]: Makes sense, done.

**Comment by mcvsubbu** [comment]
Any reason to change the name? It _is_ size based threshold updater. FYI, we have time based and row based as well.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.pinot.common.metadata.segment.SegmentZKMetadata;
+import org.apache.pinot.common.utils.LLCSegmentName;
+import org.apache.pinot.spi.stream.PartitionGroupMetadata;
+import org.apache.pinot.spi.stream.PartitionLevelStreamConfig;
+import org.apache.pinot.spi.utils.TimeUtils;
+
+public class SegmentFlushThresholdComputer {
+  public static final int MINIMUM_NUM_ROWS_THRESHOLD = 10_000;
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java:31]: Doesnt need to be public, right?

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+  // num rows to segment size ratio of last committed segment for this table
+  private double _latestSegmentRowsToSizeRatio;
+  private final Clock _clock;
+
+  public SegmentFlushThresholdComputer() {
+    this(Clock.systemUTC(), 0);
+  }
+
+  public SegmentFlushThresholdComputer(Clock clock, double latestSegmentRowsToSizeRatio) {
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java:44]: please annotate this constructor and visble for testing

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java (hunk 3 above)
Comment (mneedham) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java:32]: True, have updated

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java (hunk 4 above)
Comment (mneedham) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java]: done!

File: pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputerTest.java [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+    computer.computeThreshold(streamConfig, committingSegmentDescriptor, committingSegmentZKMetadata,
+        partitionGroupMetadataList, "events3__0__0__20211222T1646Z");
+
+    // (0.1 * (totalDocs / segmentSize)) + (0.9 * lastRatio)
+    // (0.1 * (50000 / 200000)) + (0.9 * 0.15)
+    // (0.1 * 0.25) + (0.9 * 0.15)
+    assertEquals(computer.getLatestSegmentRowsToSizeRatio(), 0.16);
+  }
+}
```
Comment (npawar) [inline pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputerTest.java:314]: nit: newline at end of file (there's an intellij setting which adds this automatically)

**Review by npawar — APPROVED** [review]
this is great, thanks Mark! lgtm

**Pull request merged**